### ISSUE TITLE
Going through geometry.py

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#137]
I'm trying to figure out why proposals from the geometry engine are so wacky, so I'm just doing a read-through to see if I can understand what is going on.

There are two very small typos with big effects that I've noted so far: We should be drawing [bonds](https://github.com/choderalab/perses/blob/master/perses/rjmc/geometry.py#L541) and [angles](https://github.com/choderalab/perses/blob/master/perses/rjmc/geometry.py#L551) from a Gaussian distribution, but [`numpy.random.random()`](http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.random.random.html) actually samples from a uniform distribution.

I'll go ahead and fix these.

**Comment by jchodera** [comment]
The other problem seems to be that instead of setting `NaN` energies to `Inf`, we were setting the `logq` to `Inf`, which is totally the opposite of what we wanted.  Fixing this too.

**Comment by jchodera** [comment]
OK, I'm seeing issues where the first few torsion proposals in `CC => CCCCC` have all `NaN`s for the torsion PMF, so there is probably something more fundamentally wrong here.

**Comment by jchodera** [comment]
I think the `NaN` issue might be numerical instability in the `Reference` platform.  Even if I set `growth_stage` to `0` or `-1`---which should turn off *everything*---I get `NaN`s.

**Comment by jchodera** [comment]
Here's the simplest demonstration script:
```python
#!/usr/bin/env python

from simtk import openmm, unit

system = openmm.System()

nparticles = 4
positions = list()
for index in range(nparticles):
    system.addParticle(1.0)
    positions.append([0,0,0])

force = openmm.CustomAngleForce('step(global_index - angle_index)*(K/2)*(theta-theta0)^2;')
force.addPerAngleParameter("theta0")
force.addPerAngleParameter("K")
force.addPerAngleParameter("angle_index")
force.addGlobalParameter('global_index', 0)
force.addAngle(0, 1, 2, [1.08, 500, 1])
system.addForce(force)

integrator = openmm.VerletIntegrator(1.0 * unit.femtoseconds)
platform = openmm.Platform.getPlatformByName('Reference')
context = openmm.Context(system, integrator, platform)
context.setPositions(positions)
print context.getState(getEnergy=True).getPotentialEnergy()
```
Output:
```
[LSKI1497:choderalab/perses/perses.choderalab] choderaj% python openmm-nans.py
nan kJ/mol
```

**Comment by jchodera** [comment]
cc https://github.com/pandegroup/openmm/issues/1397

**Comment by jchodera** [comment]
For now, I've worked around this by assigning random non-zero positions to all atoms that haven't yet been placed.

**Comment by jchodera** [comment]
Implemented the more robust fix suggested in https://github.com/pandegroup/openmm/issues/1397

**Comment by jchodera** [comment]
OK, here is why the proposals are never accepted:
```python
logp_proposal += logp_proposal + logp_r + logp_theta + logp_phi + np.log(detJ)
```
I realize you want to *really* make sure that things are added up, but I suspect that using both `+=` and including `logp_proposal` on the RHS may be a bug.  :)

**Comment by jchodera** [comment]
Something else may be wrong with the geometry placement, since the potential jumps drastically after atoms are placed:
```
potential after deletion   :       13.319 kT
potential before insertion :      115.657 kT
```

**Comment by jchodera** [comment]
This was because I was an idiot and left some debug code in.

**Comment by jchodera** [comment]
I'm seeing some crazy `logp_reverse` values now:
```
--------------------------------------------------------------------------------
Expanded Ensemble sampler iteration       25
................................................................................
MCMC sampler iteration 25
Taking 500 steps of GHMC...
Final energy is       12.449 kT
................................................................................
Updating chemical state with ncmc-geometry-ncmc scheme...
Proposing new topology...
Proposed transformation: CCC => CC
Generating new system...
Performing NCMC annihilation
Geometry engine proposal...
Performing NCMC insertion
potential after deletion   :       12.449 kT
potential before insertion :        9.664 kT
---------------------------------------------------------
switch_logp                :        2.786 kT
geometry_logp_propose      :       -1.976
geometry_logp_reverse      :    -5278.121
```

**Comment by jchodera** [comment]
This large `geometry_logp_reverse` effect seems to be due to the `logp_phi` for the last couple of torsion angles placed in the backwards direction.

**Comment by jchodera** [comment]
I changed the `_torsion_scan()` function to generate `phis` in the `[-pi,+pi)` domain to be consistent with the internal coordinate computation code, since this was causing problems otherwise.

**Comment by jchodera** [comment]
There's still something wrong with the `logp_reverse` computation, but I'm going to give up on this for now.  Current changes (all of which are bugfixes) will be merged in #136.

**Comment by jchodera** [comment]
Here's an example of the `logp_reverse` issue, where it is doing something very weird in trying to figure out how to add three more carbons and a bunch of hydrogens on that should be in some reasonable geometry:
```
Updating chemical state with ncmc-geometry-ncmc scheme...
Proposing new topology...
Proposed transformation: CCCCC => CC
Generating new system...
Performing NCMC annihilation
Geometry engine proposal...
       4 logp_r        2.529 | logp_theta        1.701 | logp_phi        1.419 | log(detJ)       -3.813
       2 logp_r    -1009.637 | logp_theta      -79.119 | logp_phi       -2.557 | log(detJ)       -2.582
      15 logp_r    -1559.048 | logp_theta      -81.133 | logp_phi     -161.158 | log(detJ)       -2.832
      16 logp_r    -3092.875 | logp_theta     -175.212 | logp_phi      -69.670 | log(detJ)       -3.052
       0 logp_r    -3230.538 | logp_theta      -36.730 | logp_phi       -1.564 | log(detJ)       -1.834
      11 logp_r    -3101.592 | logp_theta      -81.886 | logp_phi       -0.354 | log(detJ)       -2.394
      12 logp_r    -5770.680 | logp_theta      -92.826 | logp_phi      -10.319 | log(detJ)       -2.008
       5 logp_r   -28964.957 | logp_theta     -155.653 | logp_phi       -1.622 | log(detJ)       -1.123
       6 logp_r   -26937.793 | logp_theta     -208.225 | logp_phi       -8.905 | log(detJ)       -1.740
       7 logp_r   -20890.751 | logp_theta     -172.182 | logp_phi      -50.297 | log(detJ)       -1.549
```

**Comment by pgrinaway** [comment]
Ok, I think I found the problem. In the torsion scan, the values of the coordinates in `old_positions` are changed by reference, which messes up all the future logp calculations (hence the first one being ok). Using `copy.deepcopy(old_positions)` makes those ridiculous logps go away.

**Comment by pgrinaway** [comment]
Resolved in #93

**Comment by pgrinaway** [comment]
And #127
